go/teams/service_helper.go

```diff
@@ -982,49 +982,149 @@ func MemberRoleFromID(ctx context.Context, g *libkb.GlobalContext, teamID keybas
+) (res keybase1.TeamRemoveMembersResult, err error) {
+	mctx := libkb.NewMetaContext(ctx, g)
 
-		if lenExclActions > len(user.Username) && lenExclActions > len(user.Email) && lenExclActions > len(user.InviteID) {
-			g.Log.CDebugf(ctx, "RemoveMembers: can only do 1 of [username: %s, email: %s, inviteID: %s] at a time", user.Username, user.Email, user.InviteID)
-			failedToRemove = append(failedToRemove, user)
-			continue
+	// Preliminary checks
+	for _, member := range members {
+		typ, err := member.Type()
+		if err != nil {
+			return res, err
+		}
+		switch typ {
+		case keybase1.TeamMemberToRemoveType_ASSERTION:
+		case keybase1.TeamMemberToRemoveType_INVITEID:
+		default:
+			return res, fmt.Errorf("unknown TeamMemberToRemoveType %v", typ)
 		}
+	}
 
-		if len(user.Email) > 0 {
-			g.Log.CDebugf(ctx, "RemoveMembers: received email address, using CancelEmailInvite for %q in team %q", user.Email, teamID)
-			err = CancelEmailInvite(ctx, g, teamID, user.Email, user.AllowInaction)
-			if err != nil {
-				failedToRemove = append(failedToRemove, user)
+	// Removals
+	teamGetter := func() (*Team, error) {
+		return GetForTeamManagementByTeamID(ctx, g, teamID, false)
+	}
+	errstrp := func(e error) *string {
+		if e == nil {
+			return nil
+		}
+		s := e.Error()
+		return &s
+	}
+	var failures []keybase1.RemoveTeamMemberFailure
+	for _, member := range members {
+		typ, _ := member.Type()
+		switch typ {
+		case keybase1.TeamMemberToRemoveType_ASSERTION:
+			targetErr := remove(ctx, g, teamGetter, member.Assertion().Assertion)
+			var subtreeErr error
+			if member.Assertion().RemoveFromSubtree {
+				if targetErr == nil {
+					subtreeErr = removeMemberFromSubtree(mctx, teamID, member.Assertion().Assertion)
+				} else {
+					subtreeErr = errors.New("did not attempt to remove from subtree since removal failed at specified team")
+				}
 			}
-			continue
-		} else if len(user.InviteID) > 0 {
-			g.Log.CDebugf(ctx, "RemoveMembers: received inviteID, using CancelInviteByID for %q in team %q", user.InviteID, teamID)
-			err = CancelInviteByID(ctx, g, teamID, user.InviteID, user.AllowInaction)
-			if err != nil {
-				failedToRemove = append(failedToRemove, user)
+			if targetErr != nil || subtreeErr != nil {
+				failures = append(failures, keybase1.RemoveTeamMemberFailure{
+					TeamMember:     member,
+					ErrorAtTarget:  errstrp(targetErr),
+					ErrorAtSubtree: errstrp(subtreeErr),
+				})
+			}
+		case keybase1.TeamMemberToRemoveType_INVITEID:
+			targetErr := CancelInviteByID(ctx, g, teamID, member.Inviteid().InviteID)
+			if targetErr != nil {
+				failures = append(failures, keybase1.RemoveTeamMemberFailure{
+					TeamMember:     member,
+					ErrorAtSubtree: errstrp(targetErr),
+				})
 			}
+		}
+	}
+
+	res = keybase1.TeamRemoveMembersResult{Failures: failures}
+	if !shouldNotErrorOnPartialFailure && len(res.Failures) > 0 {
+		err = fmt.Errorf("failed to remove %d members", len(res.Failures))
```

Comment:
Yeah this is not ideal. res.Failures doesn't have real errors either, just strings (since it's going through AVDL). Plus there are also two possible errors "per failure.". I could template the first failure in here as a string, what do you think?

Reply:
I made a ticket.